[Issue]: optional.h uses non-member function with `->`

### Problem Description

A recent clang change (https://github.com/llvm/llvm-project/pull/90152) revealed a bug in rocThrust develop branch

https://github.com/ROCm/rocThrust/blob/f3a28e43355b0f439fb99a2210bd497ca59c8003/thrust/optional.h#L2756

rocThrust/thrust/../thrust/optional.h:2756:11: error: no member named 'construct' in 'optional<type-parameter-0-0 &>'

In `this->construct`, the arrow operator cannot be overloaded, therefore it can only access its own member function. However, construct is not its member function, since `optional<T &>` does not define `construct`. Although `optional<T>` inherits `construct` from its base classes, its member `construct` is not inherited by `optional<T &>` since they do not have inherit relation.

Since  `optional<T &>` behaves like a delegate of its member `m_value`, my guess of intention of `this->construct` is to 
perform `m_value->construct`, i.e., using the overloaded arrow operator. To make the arrow operator taking effect, `(*this)->` can be used. Therefor, the fix is

```
(*this)->construct(std::forward<Args>(args)...);
```

### Operating System

Ubuntu 22.04

### CPU

any

### GPU

AMD Radeon Pro W7900

### ROCm Version

ROCm 6.1.0

### ROCm Component

_No response_

### Steps to Reproduce

_No response_

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,9 @@ Documentation for rocThrust available at
 * Updated internal use of custom iterator in `thrust::detail::unique_by_key` to use rocPRIM's `rocprim::unique_by_key`.
 * Updated `adjecent_difference` to make use of `rocprim:adjecent_difference` when iterators are comparable and not equal otherwise use `rocprim:adjacent_difference_inplace`.
 
+### Fixes
+* Fixed incorrect implementation of `thrust::optional<T&>::emplace()`.
+
 ### Known issues
 * `thrust::reduce_by_key` outputs are not bit-wise reproducible, as run-to-run results for pseudo-associative reduction operators (e.g. floating-point arithmetic operators) are not deterministic on the same device.
 
@@ -47,7 +50,7 @@ Documentation for rocThrust available at
 
 ## rocThrust 2.18.0 for ROCm 5.7
 
-### Fixes 
+### Fixes
 * `lower_bound`, `upper_bound`, and `binary_search` failed to compile for certain types.
 * Fixed issue where `transform_iterator` would not compile with `__device__`-only operators.
 

--- a/test/test_optional.cpp
+++ b/test/test_optional.cpp
@@ -28,3 +28,41 @@ TEST(OptionalTests, IsTriviallyCopyable)
     static_assert(std::is_trivially_copyable<thrust::optional<uint64_t>>::value == true,
                   "type is not trivially copyable even though it should be!");
 }
+
+TEST(OptionalTests, EmplaceReference)
+{
+    // See https://github.com/ROCm/rocThrust/issues/404
+    {
+        int a = 10;
+
+        thrust::optional<int&> maybe(a);
+
+        int b = 20;
+        maybe.emplace(b);
+
+        ASSERT_EQ(maybe.value(), 20);
+        // Emplacing with b shouldn't change a
+        ASSERT_EQ(a, 10);
+
+        int c = 30;
+        maybe.emplace(c);
+
+        ASSERT_EQ(maybe.value(), 30);
+        ASSERT_EQ(b, 20);
+    }
+
+    {
+        thrust::optional<int&> maybe;
+
+        int b = 21;
+        maybe.emplace(b);
+
+        ASSERT_EQ(maybe.value(), 21);
+
+        int c = 31;
+        maybe.emplace(c);
+
+        ASSERT_EQ(maybe.value(), 31);
+        ASSERT_EQ(b, 21);
+    }
+}

--- a/thrust/optional.h
+++ b/thrust/optional.h
@@ -2746,14 +2746,11 @@ public:
   ///
   /// \group emplace
   __thrust_exec_check_disable__
-  template <class... Args>
+  template <class U>
   __host__ __device__
-  T &emplace(Args &&... args) noexcept {
-    static_assert(std::is_constructible<T, Args &&...>::value,
-                  "T must be constructible with Args");
-
-    *this = nullopt;
-    this->construct(std::forward<Args>(args)...);
+  T &emplace(U& u) noexcept {
+    m_value = addressof(u);
+    return *m_value;
   }
 
   /// Swaps this optional with the other.

